src/hotspot/share/cds/archiveHeapWriter.cpp

```diff
@@ -0,0 +1,658 @@
+  // into during run time, increasing the potential of memory sharing.
+  if (src_obj != nullptr) {
+    int src_hash = src_obj->identity_hash();
+    fake_oop->set_mark(markWord::prototype().copy_set_hash(src_hash));
+    assert(fake_oop->mark().is_unlocked(), "sanity");
+
+    DEBUG_ONLY(int archived_hash = fake_oop->identity_hash());
+    assert(src_hash == archived_hash, "Different hash codes: original %x, archived %x", src_hash, archived_hash);
+  }
+}
+
+// Relocate an element in the buffered copy of HeapShared::roots()
+template <typename T> void ArchiveHeapWriter::relocate_root_at(oop requested_roots, int index) {
+  size_t offset = (size_t)((objArrayOop)requested_roots)->obj_at_offset<T>(index);
+  relocate_field_in_requested_obj<T>(requested_roots, offset);
+}
+
+class ArchiveHeapWriter::EmbeddedOopRelocator: public BasicOopIterateClosure {
+  oop _src_obj;
+  oop _requested_obj;
+public:
+  EmbeddedOopRelocator(oop src_obj, oop requested_obj) :
+    _src_obj(src_obj), _requested_obj(requested_obj) {}
+
+  void do_oop(narrowOop *p) { EmbeddedOopRelocator::do_oop_work(p); }
+  void do_oop(      oop *p) { EmbeddedOopRelocator::do_oop_work(p); }
+
+private:
+  template <class T> void do_oop_work(T *p) {
+    size_t field_offset = pointer_delta(p, _src_obj, sizeof(char));
+    ArchiveHeapWriter::relocate_field_in_requested_obj<T>(_requested_obj, field_offset);
+  }
+};
+
+// Update all oop fields embedded in the buffered objects
+void ArchiveHeapWriter::relocate_embedded_oops(GrowableArrayCHeap<oop, mtClassShared>* roots,
+                                               GrowableArray<ArchiveHeapBitmapInfo>* closed_bitmaps,
+                                               GrowableArray<ArchiveHeapBitmapInfo>* open_bitmaps) {
+  size_t oopmap_unit = (UseCompressedOops ? sizeof(narrowOop) : sizeof(oop));
+  size_t closed_region_byte_size = _closed_top - _closed_bottom;
+  size_t open_region_byte_size   = _open_top   - _open_bottom;
+  ResourceBitMap closed_oopmap(closed_region_byte_size / oopmap_unit);
+  ResourceBitMap open_oopmap  (open_region_byte_size   / oopmap_unit);
+
+  _closed_oopmap = &closed_oopmap;
+  _open_oopmap = &open_oopmap;
+
+  auto iterator = [&] (oop src_obj, HeapShared::CachedOopInfo& info) {
+    oop requested_obj = requested_obj_from_buffer_offset(info.buffer_offset());
+    update_header_for_requested_obj(requested_obj, src_obj, src_obj->klass());
+    EmbeddedOopRelocator relocator(src_obj, requested_obj);
+
+    src_obj->oop_iterate(&relocator);
+  };
+  HeapShared::archived_object_cache()->iterate_all(iterator);
+
+  oop requested_roots = requested_obj_from_buffer_offset(_heap_roots_bottom);
+  update_header_for_requested_obj(requested_roots, nullptr, Universe::objectArrayKlassObj());
+  int length = roots != nullptr ? roots->length() : 0;
+  for (int i = 0; i < length; i++) {
+    if (UseCompressedOops) {
+      relocate_root_at<narrowOop>(requested_roots, i);
```

Comment:
This can also follow similar pattern as code in my comment on`do_oop_work` above, as in:

```
    if (UseCompressedOops) {
      offset = (size_t)((objArrayOop)requested_roots)->obj_at_offset<narrowOop>(i);
      narrowOop* buffered_addr = (narrowOop*)(cast_from_oop<address>(buffered_roots) + offset);
      oop request_referent = source_obj_to_requested_obj(CompressedOops::decode(*buffered_addr));
      *buffered_addr = CompressedOops::encode(request_referent);
      mark_oop_pointer<narrowOop>(requested_roots, offset, heap_info->oopmap());
    }
```

Reply:
In light of the recent change, this has also been taken care of.